### PR TITLE
Initial eslint custom rule / plugin

### PR DESCRIPTION
**Pull request by chuck-dbos**
Assuming there is agreement about a top-level repo and package, the follow-on work is:

* Add a base set of rules for extension
* Add usage to documentation and examples

### CODE REVIEW

**Review by kraftp — APPROVED** [review]
Looks good and the rules make sense!

I don't fully understand how developers will use it.  Do these work like any other eslint rules, so you add them to your `.eslintrc`?  Will we upload this to npm as its own package to make that work?

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+# Optional REPL history
+.node_repl_history
+
+# Output of 'npm pack'
+*.tgz
+
+# Editor temp files
+*.swp
+*.swo
+examples/hello/.env
```
Comment (maxdml) [inline .gitignore:38]: is this `.env` file created by VSCode?

File: dbos-rules.test.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  ruleUnderTest.rules['detect-nondeterministic-calls'], // rule code
+  { // checks
+    // 'valid' checks cases that should pass
+    valid: [{
+      code: "const foo = 'bar';",
+    }],
+    // 'invalid' checks cases that should not pass
+    invalid: [{
+      code: "const foo = Math.random();",
+      //output: 'const foo = *NEED SUGGESTION*;',
```
Comment (maxdml) [inline dbos-rules.test.js:20]: You mean: what do we set as an error message for the lint rule?

File: dbos-rules.js [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+          },
+        };
+      },
+    },
+    'detect-new-date': {
+      // Rule configuration for new Date() detection
+      meta: {
+        type: 'suggestion',
+        docs: {
+          description: 'Detect calls to new Date(), which should be called via DBOS rather than directly',
```
Comment (maxdml) [inline dbos-rules.js:98]: Naïve question: what does "called via DBOS" means?

**Comment by chuck-dbos** [comment]
> Looks good and the rules make sense!
> 
> I don't fully understand how developers will use it. Do these work like any other eslint rules, so you add them to your `.eslintrc`? Will we upload this to npm as its own package to make that work?

It gets a lot easier as a published package, yes.  All you will do is install `eslint`, `@dbos-inc/eslint-plugin`, add a command to your `package.json`, and add `.eslintrc` (or a dreadful flat config):
```
  "extends": [
    "plugin:@dbos-inc/dbosRecommendedConfig"
  ],
  "plugins": [
    "@dbos-inc",
  ],
```

The directory initializer can certainly do this for you, and it can be placed in all examples, quick-start guide, etc.

The example code - with the stuff to find - is currently in the demo apps on a branch.  I will split that up a bit soon to separate it from the compiler API part.

File: dbos-rules.js (hunk 3 above)
Comment (chuck-dbos) [reply dbos-rules.js:98]: > Naïve question: what does "called via DBOS" means?

This is not final.   I mentioned in a meeting that we ought to offer some function to call - the context, a communicator library, or something we could do in the debugger to override the return value... but I didn't get buy-in on any proposals.

File: .gitignore (hunk 1 above)
Comment (chuck-dbos) [reply .gitignore:38]: I don't know... I am sure I cut+paste it.

**Pull request merged**